# Nextflow RNA-seq pipline
![pipline diagram](diagram.svg)
## Introduction
This pipeline was created as part of a project of the course "Computational Workflows". It aims to provide minimal functionality for the analysis of RNA-seq data, while being reproducible and modular. The pipeline is inspired and uses code from the [nf-core/rnaseq](https://nf-co.re/rnaseq/3.16.0/) pipeline (version 3.16.0).

#### Steps

1. The [FastQC](https://nf-co.re/modules/fastqc) module of nf-core is used to generate quality reports of the input reads.

2. The nf-core module for [Trim Galore](https://nf-co.re/modules/trimgalore) is used to trim the raw reads.

3. [HISAT2-build](https://nf-co.re/modules/hisat2_build) and [HISAT2-align](https://nf-co.re/modules/hisat2_align) are used to index the reference genome (.fasta and .gtf file) and align the reads to the reference.

4. The [SAMtools-sort](https://nf-co.re/modules/samtools_sort) module is used to sort the output alignment (.bam) of HISAT2. This step is required for subsequent feature counting.

5. The nf-core [subreads-featurecounts](https://nf-co.re/modules/subread_featurecounts) module is used to count the alignments per genomic feature.

## Usage

The pipeline can be used with the following parameters:

```bash
nextflow run main.nf\
    --input <PATH_TO_SAMPLESHEET.csv> \
    --fasta <PATH_TO_REFERENCE_GENOME.fa> \
    --gtf <PATH_TO_GTF.gtf> \
    --outdir <PATH_TO_OUT_DIRECTORY> \
    --threads <#_OF_THREADS> \
    --ram <#_OF_RAM_GB> \
    -profile <conda/docker> \
```

`--input` path to the __samplesheet__. An exemplary [samplesheet](tests/samplesheet_test.csv) can be found in the [tests](tests/) folder and should follow the following format:

```
sample,fastq_1,fastq_2,strandedness
SRR23195516_oxy_sni,./tests/data/reduced_SRX19144486_SRR23195516_1.fastq.gz,./tests/data/reduced_SRX19144486_SRR23195516_2.fastq.gz,auto
SRR23195511_oxy_sham,./tests/data/reduced_SRX19144488_SRR23195511_1.fastq.gz,./tests/data/reduced_SRX19144488_SRR23195511_2.fastq.gz,auto
```

`--fasta` path to the reference genome FASTA file.

`--gtf` path to the reference genomes feature file.

`--outdir` path to a folder where all the outputs should be stored.

`--threads` maximum number of threads to be used for the computation

`--ram` maximum amount of RAM to use during computation

`-profile` profile to run the pipeline with.

#### Example run
Exemplary command to run the pipeline with the included test files in the [tests](tests/) directory. The reference genome files (fasta and gtf) had to be excluded from the github repo, since they are too large. The files can be downloaded from [NCBI](https://www.ncbi.nlm.nih.gov/datasets/genome/GCF_000001635.27/) and have to be extracted.

```bash
nextflow run main.nf\
    --input ./tests/samplesheet_test.csv \
    --fasta ./tests/data/GCF_000001635.27_GRCm39_genomic.fna \
    --gtf ./tests/data/GCF_000001635.27_GRCm39_genomic.gtf \
    --outdir ./tests/out_test/ \
    --threads 12 \
    --ram 8GB \
    -profile docker \
```

## Pipline Output

The output of all steps of the pipeline can be found in the folder specified with `--outdir`. Exemplary output files can be found in the [tests/out_test/](tests/out_test/) folder.

* FastQC generates quality reports in HTML format ([example](tests/out_test/fastqc/SRR23195511_oxy_sham_1_fastqc.html))
* Trim Galore generates .fasta files with the trimmed reads, as well as trimming report files ([example](tests/out_test/trimgalore/)).
* The HISAT2 alignments and summary files are stored in the [hisat2](tests/out_test/hisat2/) folder
* The sorted alignment files as generated by SAMtools can be found in the [samtools](tests/out_test/samtools/) folder
* The feature counts from Subread featureCounts can be accessed in the [counts](tests/out_test/counts/) folder
* The versions of all tools that were used over the nf-core modules can be found in the [versions/versions.yml](tests/out_test/versions/versions.yml) file